### DISCUSSION TITLE
Virtuak ZPL Printer - Browser Z** Print

### DISCUSSION THREAD

**Discussion by CMIGIT** [porrey/Virtual-ZPL-Printer#56] [General]
I tried to make Virtual ZPL Printer work with Z* Browser Print.
On my PC it didn't work, and I found that it was caused by an exception during the stream reading.

I replaced in TcpListenerClientHandler (near line 90)

```
while ((numBytesRead = stream.Read(data, 0, data.Length)) > 0)
{
		ms.Write(data, 0, numBytesRead);
}
```

by

```
while ((numBytesRead = stream.Read(data, 0, data.Length)) > 0)
{
		ms.Write(data, 0, numBytesRead);
		if (numBytesRead < data.Length)
		{
				break;
		}
}
```

And it seems to work better. But not yet 100%.
Not all labels come in Virtual ZPL Printer, and sometimes I have errors between my web site and Z* Browser Print.
(no error in Virtual ZPL Printer)

(Sorry I'm not familiar enough with Git to make pull requests.)

I don't know if I'm the only one having problems with this configuration with Z* Browser Print.

**Comment by porrey** [comment]
I downloaded **Zebra Browser Print** and tested it. I had to make a few minor changes to support it.

I found that the app does not always send a request to the printer. About every third button push works. When it does NOT work, I do not get a request from the app, but when a request is received, it works every time.

In the next release, there will be a few changes to help support **Zebra Browser Print**.

**Comment by porrey** [comment]
Updates have been released.

**Comment by porrey** [comment]
Please let me know if you have had a chance to test again.

**Comment by CMIGIT** [comment]
When Zebra browser print send data, all is ok ;-)
I'll make other tests with real printer, if I'll see somthing... 
I don't know if ZBP is waiting for a "ack"

**Comment by CMIGIT** [comment]
I'm back on the subject.

Ok I think I finally find how Browser Print works.
Borwser Print is open a connection and send label, but keep the connection open, and sends new label on this connection.
So in TcpListenerClientHandler, you need to loop and continue to read data with "networkStream.ReadAsync(data);" (no directly close the connection) until "if (client.Connected && networkStream.CanRead)" (with Tread.Sleep of course)

I hope it can help to the communication betwwen "Borwser Print" and "Virtual ZPL Printer" , because I think it can be a good combination for work.

**Comment by CMIGIT** [comment]
No so easy, but I have made "some" update in "TcpListenerClientHandler.cs".
I Join the file
[TcpListenerClientHandler.cs.txt](https://github.com/user-attachments/files/17795840/TcpListenerClientHandler.cs.txt)
